pyartcd/pyartcd/jenkins.py

```diff
@@ -1,74 +1,161 @@
+    If 'BUILD_URL' env var was provided, add a TRIGGERED_FROM
+    field to the build params to allow backlinking to the original job
     """
 
-    service_account = os.environ['JENKINS_SERVICE_ACCOUNT']
-    token = os.environ['JENKINS_SERVICE_ACCOUNT_TOKEN']
+    try:
+        build_url = os.environ['BUILD_URL']
+    except KeyError:
+        logger.error('A BUILD_URL env var must be defined!')
+        raise
 
-    build_url = os.environ.get('BUILD_URL', '')
     if build_url:
         params['TRIGGERED_FROM'] = build_url
+    return params
+
+
+def block_until_building(queue_item: QueueItem, delay: int = 5) -> str:
+    """
+    Watches a queue item and blocks until the scheduled build starts.
+
+    Returns the URL of the new build.
+    """
+
+    while True:
+        try:
+            data: dict = queue_item.poll()
+            build_number = data['executable']['number']
+            break
+        except (KeyError, TypeError):
+            logger.info('Build not started yet, sleeping for %s seconds...', delay)
+            time.sleep(delay)
+
+    logger.info('Build started: number = %s', build_number)
+    return f"{data['task']['url']}{build_number}"
+
+
+def block_until_complete(queue_item: QueueItem, delay: int = 60) -> str:
+    """
+    Watches a queue item and blocks until the scheduled build completes
+
+    Returns the build result as a string
+    """
+
+    while True:
+        try:
+            data: dict = queue_item.poll()
+            build_number = data['result']
+            break
+        except (KeyError, TypeError):
+            logger.info('Build not completed yet, sleeping for %s seconds...', delay)
+            time.sleep(delay)
+
+    logger.info('Build started: number = %s', build_number)
+    return data['url']
+
+
+def start_build(job_name: str, params: dict, blocking: bool = False,
+                watch_building_delay: int = 5, watch_complete_delay: int = 60) -> str:
+    """
+    Starts a new Jenkins build
+
+    :param job_name: e.g. "aos-cd-builds/build%2Fbuild-sync"
+    :param params: a key-value collection to be passed to the build
+    :param blocking: if True, will block until the new builds completes; if False,
+                    start a new build in "fire-and-forget" fashion
+    :param watch_building_delay: Poll rate for building state
+    :param watch_complete_delay: Poll rate for completed state
+
+    Returns the build result if blocking=True, the empty string otherwise
+    """
+
+    logger.info('Starting new build for job: %s', job_name)
+    queue_item = jenkins.get_job(job_name).invoke(build_params=params)
+    build_url = block_until_building(queue_item, watch_building_delay)
+    logger.info('Started new build at %s', build_url)
+
+    # If blocking==True, wait for the build to complete; get its status and return it
+    if blocking:
+        logger.info('Waiting for build to complete...')
+        build_result = block_until_complete(queue_item, watch_complete_delay)
```

Comment:
since we already have the build number, we should use build.is_building() to check the status, the queue item will be destroyed after it triggers the build.

Reply:
Right. Fixed in 5b6922ee3814d47449511d6e3018e70d4bee3c4d

Test build: https://saml.buildvm.hosts.prod.psi.bos.redhat.com:8888/job/hack/job/dpaolell/job/test-remote/70/console